src/libcore/mem/manually_drop.rs

```diff
@@ -113,19 +123,28 @@ impl<T> ManuallyDrop<T> {
 }
 
 impl<T: ?Sized> ManuallyDrop<T> {
-    /// Manually drops the contained value.
+    /// Manually drops the contained value. This is exactly equivalent to calling
+    /// [`ptr::drop_in_place`] with a pointer to the contained value. As such, unless
+    /// the contained value is a packed struct, the destructor will be called in-place
+    /// without moving the value, and this can be used to safely drop [pinned] data.
     ///
     /// If you have ownership of the value, you can use [`ManuallyDrop::into_inner`] instead.
     ///
     /// # Safety
     ///
-    /// This function runs the destructor of the contained value and thus the wrapped value
-    /// now represents uninitialized data. It is up to the user of this method to ensure the
-    /// uninitialized data is not actually used.
-    /// In particular, this function can only be called at most once
-    /// for a given instance of `ManuallyDrop<T>`.
+    /// This function runs the destructor of the contained value. Other than changes made by
+    /// the destructor itself, the memory is left unchanged, and so as far as the compiler is
+    /// concerned still holds a bit-pattern which is valid for the type `T`.
+    ///
+    /// However, this "zombie" value should not be exposed to safe code, and this function
+    /// should not be called more than once. To use a value after it's been dropped, or drop
+    /// a value multiple times, can cause Undefined Behavior (depending on what `drop` does).
+    /// This is normally prevented by the type-system, but users of `ManuallyDrop` must
+    /// uphold those guarantees without assistance from the compiler.
```

Comment:
Maybe "should" is a bit weak here, and we should use "must"?

Reply:
There are still cases where it might be valid: for example, if the type is `Copy`, or if the destructor is within the "unsafe code boundary" and is known to have well-defined behaviour if called multiple times, but I didn't want to go too far into the details.

Reply:
Hm... fair. Though:

>  if the destructor is within the "unsafe code boundary" and is known to have well-defined behaviour if called multiple times

That is not guaranteed, destructors might have "magic extra semantics" that makes calling them multiple times UB no matter what they do.